src/vs/code/node/cli.ts

```diff
@@ -80,46 +79,78 @@ export async function main(argv: string[]): TPromise<any> {
-			try {
-				const stdinFileStream = fs.createWriteStream(stdinFilePath);
-				resolveTerminalEncoding(verbose).done(encoding => {
-
-					// Pipe into tmp file using terminals encoding
-					const converterStream = iconv.decodeStream(encoding);
-					process.stdin.pipe(converterStream).pipe(stdinFileStream);
-				});
-
-				// Make sure to open tmp file
-				argv.push(stdinFilePath);
+		if (stdinWithoutTty) {
+
+			// Read from stdin: we require a single "-" argument to be passed in order to start reading from
+			// stdin. We do this because there is no reliable way to find out if data is piped to stdin. Just
+			// checking for stdin being connected to a TTY is not enough (https://github.com/Microsoft/vscode/issues/40351)
+			if (args._.length === 1 && args._[0] === '-') {
+
+				// remove the "-" argument when we read from stdin
+				args._ = [];
+				argv = argv.filter(a => a !== '-');
+
+				// prepare temp file to read stdin to
+				stdinFilePath = paths.join(os.tmpdir(), `code-stdin-${Math.random().toString(36).replace(/[^a-z]+/g, '').substr(0, 3)}.txt`);
+				let stdinFileError: Error;
+				try {
+					const stdinFileStream = fs.createWriteStream(stdinFilePath);
+					resolveTerminalEncoding(verbose).done(encoding => {
+
+						// Pipe into tmp file using terminals encoding
+						const converterStream = iconv.decodeStream(encoding);
+						process.stdin.pipe(converterStream).pipe(stdinFileStream);
+					});
+
+					// Make sure to open tmp file
+					argv.push(stdinFilePath);
+
+					// Enable --wait to get all data and ignore adding this to history
+					argv.push('--wait');
+					argv.push('--skip-add-to-recently-opened');
+					args.wait = true;
+				} catch (error) {
+					stdinFileError = error;
+				}
 
-				// Enable --wait to get all data and ignore adding this to history
-				argv.push('--wait');
-				argv.push('--skip-add-to-recently-opened');
-				args.wait = true;
-			} catch (error) {
-				stdinFileError = error;
+				if (verbose) {
+					if (stdinFileError) {
+						console.error(`Failed to create file to read via stdin: ${stdinFileError.toString()}`);
+					} else {
+						console.log(`Reading from stdin via: ${stdinFilePath}`);
+					}
+				}
 			}
 
-			if (verbose) {
-				if (stdinFileError) {
-					console.error(`Failed to create file to read via stdin: ${stdinFileError.toString()}`);
-				} else {
-					console.log(`Reading from stdin via: ${stdinFilePath}`);
-				}
+			// If the user pipes data via stdin but forgot to add the "-" argument, help by printing a message
+			// if we detect that data flows into via stdin after a certain timeout.
+			else if (args._.length === 0) {
+				processCallbacks.push(child => new TPromise(c => {
+
+					// wait for 1s maximum...
+					setTimeout(() => {
+						c(void 0);
+					}, 1000);
+
+					// ...but finish early if we detect data
+					process.stdin.on('data', () => {
```

Comment:
Should this be `once` instead of `on`?

Reply:
Also, should the listener be detached when this promise resolves?

Reply:
@joaomoreno good point, fixed via https://github.com/Microsoft/vscode/commit/aa8895a6b8c7e73584f31eefac4c85b38b93ec1b